### ISSUE TITLE
Lemmas for numbers

### ISSUE THREAD

**Issue by amir-zeldes** [UniversalDependencies/UD_English-EWT#113]
Currently lemmatization is inconsistent, at least across EWT and GUM, and I'd like to standardize practices for at least the following two cases:

  * Numbers with separators: currently "2000" and "2,000" are lemmatized to themselves in EWT, but both to "2000" in GUM, which I think I like better, since I feel like the purpose of lemmatization is to collapse forms that correspond to the same lexical item.
  * Spelled out numbers: currently "two" and "2" are lemmatized "two" and "2" in both GUM and EWT. I'd like to change both to "2". 

Basically I propose that NUM should use Arabic numerals whenever possible, which may be useful to suggest as a cross-linguistic standard to help with multilingual convergence - how would everyone feel about that?

Adding @nschneid @sebschu @ethanachi

**Comment by nschneid** [comment]
Fine with me!

**Comment by nschneid** [comment]
For English, anyway. Is there a chance that for other languages it would be worthwhile to have the spelled-out form (rather than the Arabic numeral) for morphological reasons?

Also, what about ordinals?

**Comment by amir-zeldes** [comment]
If the lemma needs to have a gender form, and generally the language distinguishes gender for the lemma, then I would agree they should not use Arabic numerals just because of this. But if nothing speaks against it, I think it would contribute to cross-linguistic convergence if there is a recommendation to use Arabic numerals if possible.

In any case, if we agree on this for English, I can go ahead and update GUM, or also EWT whenever all PRs in dev are merged.

**Comment by ethanachi** [comment]
Definitely agree on the first point (2,000 definitely feels like the variant form here). 

On the second, I share the same concern as @nschneid; does it help multilingual convergence necessarily to use the same lemmas for numbers cross-lingually?

**Comment by amir-zeldes** [comment]
OK, so it sounds like we're reaching consensus on 1. Can we agree on 2 for English? If so, then we don't need to tackle whether it's a good universal guideline yet. I'd also defer ordinals until we agree on this one (there are no cross-lingual Arabic ordinal forms anyway, so that would definitely only be for English at first)

**Comment by nschneid** [comment]
If we're not sure it would be a good guideline crosslinguistically then we should think about how to encode the information via features so as not to lose anything. I see some treebanks use [`NumForm`](https://universaldependencies.org/u/feat/NumForm.html).

What about multiword number names like "two hundred"? Making the lemmas "2" and "100" respectively could be misleading to tools that look at lemmas individually.

**Comment by amir-zeldes** [comment]
+1 for NumForm, I can add that to GUM easily. 

I don't think 2 + 100 is misleading, since it's no worse than "two + hundred". Lemmas are token-level attributes, so it would be wrong to expect them to express compositional number values that come from multiple tokens. And if we wanted to extract such values from multiple tokens, having Arabic numerals would still be a better starting point than a variety of ways of writing each component number.

**Comment by nschneid** [comment]
Although technically, in edited text, one would expect the token "100" to be read as "a hundred" or "one hundred"; you wouldn't write "2 100" meaning "200". So the spelled-out and Arabic forms aren't completely interchangeable.

But I don't have a strong opinion as long as it's a consistent policy and all the info is represented somewhere between the wordform, lemma, and feats.

**Comment by amir-zeldes** [comment]
OK, in that case I will do for English:

form  | lemma  | NumForm (in FEATS)
---- | ----- | ------------------
Two | 2 | Word
1000 | 1000 | Digit
10,000 | 10000 | Digit
2.1 | 2.1 | Digit
XIV | 14 | Roman

And actually Gender is in FEATS anyway, so it's possible to combine with gendered numbers, or not, if a language has guidelines that generally enforce gendered lemmas.

**Comment by rueter** [comment]
Hi @amir-zeldes, I would like to draw your attention to at least two problems in lemmatizing Roman numerals to Arabic numerals in a real world scenario. 
(1) In books, sometimes page numeration may include a strategy where an introductory segment of an initial volume will have Roman numerals for page numeration, after which Arabic numerals commence at 1 (see H. Paasonens Mordwinisches Wörterbuch. Lexica Societatis Fenno-Ugricae XXIII. (Kotimaisten kielten tutkimuskeskuksen julkaisuja 59).  XXIII:1. 1990. CVI + 557 p. Band I (A–J).)
(2) The University of Helsinki has a main building, where lecture halls on one side of the building are listed by Arabic numerals, and lecture halls on the other side are listed by Roman numerals. The equation of Arabic and Roman numerals is, therefore, misleading and exactly what a foreign reference to my dissertation held in December of 2010 did.

**Comment by sylvainkahane** [comment]
Is it really a decision that should be specific to English?
I think it is more a problem of genres than particular languages. I would like just to mention that in spoken corpora the tradition is to transcribe all numbers by words. There is still no treebank of spoken English but we can hope that there will be one some day.

**Comment by amir-zeldes** [comment]
@rueter I definitely agree that Roman and Arabic numerals should be distinguished for the reasons you pointed out. But lemmatization is not a lossless process: it produces many ambiguities, and this would be one. The form field still distinguishes everything, while lemma can tell us that both 2 and II mean "two" in some way, and NumType can tell us about the categorical nature of the difference between them.

@sylvainkahane GUM does include some spoken English data in its interview section, and I'm happy to report that three of the four new genres coming to the corpus this winter will also be spoken!

**Comment by dan-zeman** [comment]
While I am in general for a relatively great degree of normalization in lemmas, I would keep "2" and "two" separate. If they are collapsed, I would prefer "two" because I consider the written words primary in a language (but then we will have issues with lemmas for things like "3972"). BTW do we have the same lemma for abbreviations and full words?

**Comment by nschneid** [comment]
Re: abbreviations—it may depend on the corpus; I don't think EWT expands abbreviations, and if we wanted to it would be quite a lot of work to implement. [The guidelines](https://universaldependencies.org/u/overview/morphology.html) say "Abbreviated/shortened forms can be mapped to their full spelling as the lemma in conjunction with the feature `Abbr=Yes`, provided that the full spelling is a single word. Abbreviations that would expand to multiple words should be retained in the lemma."

Re: numbers—@dan-zeman, would you prefer adding a feature for the Arabic form of the number, rather than the lemma?

**Comment by nschneid** [comment]
(A meta-point is that there are several different levels of normalization that can be applied in a text—removing inflectional morphology, typos, segmentation errors; expanding abbreviations; unifying orthographic variants—and different ones will be relevant for different purposes. So perhaps we should be conservative about putting too much normalization in the lemma field if some kinds of normalizations can go elsewhere.)

**Comment by dan-zeman** [comment]
> Re: numbers—@dan-zeman, would you prefer adding a feature for the Arabic form of the number, rather than the lemma?

It would not be a morphological feature, hence it would have to go to MISC. I already have it in Czech and the attribute is called `LNumValue`:

(_padesát marek_ = "fifty marks")

<pre>
1	Padesát	padesát	NUM	Cn-S4----------	Case=Acc|Number=Sing|NumForm=Word|NumType=Card	2	nummod:gov	2:nummod:gov	LNumValue=50
2	marek	marka	NOUN	NNFP2-----A----	Case=Gen|Gender=Fem|Number=Plur|Polarity=Pos	9	obj	9:obj	_
</pre>

In the original annotation of the Prague Dependency Treebank, it was actually part of the lemma (if I'm not mistaken, the lemma was <tt>padesát`50</tt>). In UD the lemma is supposed to be an existing form of the lexeme without any extra tags or ids (of which the PDT lemmas have plenty), so I strip all such extra info from the lemma and put it in MISC attributes.

**Comment by amir-zeldes** [comment]
@dan-zeman I think going with 'always words' is not an option for the reason you brought up - we don't want spaces in lemmas and can't lemmas like "threehundredandsixtyfive"... 

As for Arabic vs. same as form, from a practical perspective, I can't think of a situation in which I'd be sad to have the Arabic numerals in lemma (if I want the word, it's in form anyway, and `NumForm` is an elegant way to know what we're dealing with), but I can think of situations where the opposite would be less useful (need to do more work to extract numerical values). 

From a more theoretical perspective, I think that lemmas correspond to something like "dictionary entries", and "two" and "2", at least for me, are different ways of writing the same word, so they wouldn't get separate dictionary entries. But if the idea is not popular, I can easily do it one way or the other.

**Comment by dan-zeman** [comment]
@amir-zeldes I see your point and the opposite (lemma "two" + LNumValue in MISC) is just my personal preference, so it does not weigh more than one vote. I would be willing to defend my position though if you wanted to extend your approach to other languages than English. Because in English you can say "if I want the word, it's in form anyway", but this is not necessarily true outside English. In Czech, the lemma _dva_ ("two") represents four morphological forms: _dva, dvě, dvou, dvěma._

**Comment by amir-zeldes** [comment]
> In Czech, the lemma dva ("two") represents four morphological forms: dva, dvě, dvou, dvěma.

Based on the discussion above, I'm definitely restricting this issue to English at the moment. But I'm not sure I understand the Czech objection: dva, dvě, dvou, dvěma are all possible forms of the number two, in all genders, right? So what would be the harm in lemmatizing to "2" and relying on the form to know which one it was, and then using `Gender`, `Case` and `Number` to know the specific analysis? 

We can even go a step further to Polish, since it has forms that can only be a specific gender: "dwaj" can only be masculine (nominative/vocative) and "dwie" can only be feminine (nominative/vocative or accusative), while "dwa" is neuter (same cases). So in this case, an argument could be made that there should be three lemmas, one per gender:

  * dwa - forms: dwa, dwu, wdum, dwóch, dwoma
  * dwaj - forms: dwaj , dwu, wdum, dwóch, dwoma
  * dwie - forms: dwie , dwu, wdum, dwóch, dwiema

If the Polish treebanks distinguished gender in lemmas in general, then I could see an argument that it would be inconsistent for numbers not to follow the same pattern. However, Polish TBs actually lemmatize all adjectives to the masculine nominative singular, and consistent with that, also lemmatize numbers to the masculine singular as well:

```
UD Polish PDB:
# sent_id = dev-s9
1	Dwie	dwa	NUM	num:pl:nom:f:congr	Case=Nom|Gender=Fem|Number=Plur|NumForm=Word	2	nummod	2:nummod	_
2	kobiety	kobieta	NOUN	subst:pl:nom:f	Case=Nom|Gender=Fem|Number=Plur	7	nsubj	7:nsubj	_
...
```

```
UD Polish LFG
# sent_id = dev-69
1	Ale	ale	CCONJ	conj	_	5	cc	5:cc	_
2	w	w	ADP	prep:loc:nwok	AdpType=Prep|Variant=Short	4	case	4:case	Case=Loc
3	samym	sam	ADJ	adj:sg:loc:n:pos	Case=Loc|Degree=Pos|Gender=Neut|Number=Sing	4	amod	4:amod	_
4	centrum	centrum	NOUN	subst:sg:loc:n	Case=Loc|Gender=Neut|Number=Sing	5	obl	5:obl:w	_
5	zmniejszyła	zmniejszyć	VERB	praet:sg:f:perf	Aspect=Perf|Gender=Fem|Mood=Ind|Number=Sing|Tense=Past|VerbForm=Fin|Voice=Act	0	root	0:root	_
6	o	o	ADP	prep:acc	AdpType=Prep	7	case	7:case	Case=Acc
7	dwie	dwa	NUM	num:pl:acc:f:congr	Case=Acc|Gender=Fem|Number=Plur|NumType=Card	5	obl	5:obl:o	DepType=Congr
8	trzecie	trzeci	ADJ	adj:pl:acc:f:pos	Case=Acc|Degree=Pos|Gender=Fem|Number=Plur	7	amod	7:amod	SpaceAfter=No
9	.	.	PUNCT	interp	PunctType=Peri	5	punct	5:punct	_
```

So all this proposal would amount to is replacing a current consistent identifier (currently "dwa", regardless of word form/case/gender) with another ("2"). But as I said above, I'm really just asking about English now, since we're revising GUM and I want the lemma layer to be as clean as possible.

**Comment by dan-zeman** [comment]
> So what would be the harm in lemmatizing to "2" and relying on the form to know which one it was, and then using Gender, Case and Number to know the specific analysis?

Of course we use the features in addition to the lemma. The potential harm could occur if somebody wanted to model morphological paradigms by comparing the lemma with individual forms. One can see that _dvě_ is somehow inflectionally related to _dva_, but with _2_ it is lost.

**Comment by amir-zeldes** [comment]
OK, I actually just looked into Czech PDT, and if I understand it right, it's exactly the same as Polish - the masculine form is used for the lemma of all genders of "two", the same as adjectives:

```
# sent_id = cmpr9407-012-p4s1
1	BESO	Beso	PROPN	NNNS1-----A----	Case=Nom|Gender=Neut|NameType=Com|Number=Sing|Polarity=Pos	6	nsubj	6:nsubj	_
2	a	a	CCONJ	J^-------------	_	3	cc	3:cc	LId=a-1
3	SES	SES	PROPN	NNXXX-----A---8	Abbr=Yes|NameType=Com|Polarity=Pos	1	conj	1:conj|6:nsubj	_
4	jsou	být	AUX	VB-P---3P-AA---	Mood=Ind|Number=Plur|Person=3|Polarity=Pos|Tense=Pres|VerbForm=Fin|Voice=Act	6	cop	6:cop	_
5	dvě	dva	NUM	ClHP1----------	Case=Nom|Gender=Fem,Neut|Number=Plur|NumForm=Word|NumType=Card|NumValue=1,2,3	6	nummod	6:nummod	LNumValue=2
6	organizace	organizace	NOUN	NNFP1-----A----	Case=Nom|Gender=Fem|Number=Plur|Polarity=Pos	0	root	0:root	_
7	se	s	ADP	RV--7----------	AdpType=Voc|Case=Ins	9	case	9:case	LId=s-1
8	stejným	stejný	ADJ	AANS7----1A----	Case=Ins|Degree=Pos|Gender=Neut|Number=Sing|Polarity=Pos	9	amod	9:amod	_
9	posláním	poslání	NOUN	NNNS7-----A----	Case=Ins|Gender=Neut|Number=Sing|Polarity=Pos	6	nmod	6:nmod:s:ins	SpaceAfter=No|LDeriv=poslat
10	.	.	PUNCT	Z:-------------	_	6	punct	6:punct	_
```

So feminine dvě is also lemmatized to "dva", and there is no use of lemma to distinguish the gender. So as in Polish, changing to "2" would just replace one lemma with another, not merge gendered lemmas.

**Comment by amir-zeldes** [comment]
> The potential harm could occur if somebody wanted to model morphological paradigms by comparing the lemma with individual forms. One can see that dvě is somehow inflectionally related to dva, but with 2 it is lost.

I see, so is the issue more related to computational tools which could recognize systematic relations between stems and inflected forms? 

If so, I'm not sure how to weight this consideration, but I would say in any case you already have cases of the form "2", for which inflection cannot be recognized, so those are also opaque to such tools. In practice this might not be so important for numbers, since spelled out numbers are almost exclusively a small set that can be handled in a list. But if preserving string-overlap between lemma and form is a high priority, then I think you will have to have two lemmas, "2" and "dva". I think the price of that is basically saying that there are two separate lexical items for every number: its Arabic form and its spelled out form (and maybe even more, if we include Roman or other representations).

**Comment by dan-zeman** [comment]
> I think the price of that is basically saying that there are two separate lexical items for every number: its Arabic form and its spelled out form

Yes, that's what we do :-) It is similar to having separate lemmas for _$_ and _dollar_ (of course you could argue that they, too, should be collapsed to one lemma).

**Comment by nschneid** [comment]
Right, I think if we really wanted to be consistent and treat a lemma as a canonical dictionary entry, we would need to normalize nonnumeric symbols and abbreviations as well.

But maybe the practical advantages of normalizing numbers in the lemma field (for numeric reasoning applications or whatever) outweigh the linguistic purity of a consistent definition of lemma.

**Comment by amir-zeldes** [comment]
I think it's possible to have a guideline for numbers without rethinking all lemmatizations for the language :) Or other languages than English, for that matter. They say perfect is the enemy of the good...

If I can get a show of hands (or + signs) from participants who care about this issue for English, I can provide a depedit script to revise the lemmas in current UD treebanks to numbers (or if more people vote against it, we can also close the issue).

**Comment by dan-zeman** [comment]
I guess my two cents are still on the minus side. But what really matters is whether you can reach consensus among the people maintaining the 8 existing UD treebanks for English. I don't know whether all of them follow this discussion.

**Comment by amir-zeldes** [comment]
Fair enough :)

In that case let's ping the latest committers to the various TBs - I manage GUM/GUMReddit, @ethanachi and @sebschu are the last committers to EWT and PUD, @msang for ParTut, @LarsAhrenberg for LinES, @rmunro for pronouns, and @berzak for ESL. 

The question is: would you accept a pull request unifying the lemmas of numeral expressions to the corresponding Arabic numeral, with an additional feature indicating the numeral type (spelled out "two", Arabic "2", or Roman "II")? The result would be that word forms like "2", "two" and "II" would have the same lemma. 

If so, I'm willing to do the edits and submit PRs, and GUM will follow the same behavior.

**Comment by LarsAhrenberg** [comment]
Currently in LinES '2' and 'two' have different lemmas. There are no instances of Roman numbers, as far as I recall, but I think I would lemmatize them as such. I think of lemmas as representatives of a morphological/orthographical paradigms, not as semantic constants, similarly to Dan's argument above. Even in English, I believe you can write '2nd' but not 'twond' for second. So, while I consent to giving '10000' as the lemma of '10,000' I vote against the proposed radical unification of numerical expressions.

**Comment by AngledLuffa** [comment]
What about sentences with European style numbers?  Do we want to change 3,40 to 3.40 in the lemma here?

```
# sent_id = answers-20111108024148AAO8oFI_ans-0010
# text = Price:3,40 Euros, 5 Euros or 7,5 Euros (1) for a 72 heures lenght.
1       Price   price   NOUN    NN      Number=Sing     0       root    0:root  SpaceAfter=No
2       :       :       PUNCT   :       _       1       punct   1:punct SpaceAfter=No
3       3,40    3,40    NUM     CD      NumType=Card    4       nummod  4:nummod        _
4       Euros   euro    NOUN    NNS     Number=Plur     1       appos   1:appos SpaceAfter=No
5       ,       ,       PUNCT   ,       _       7       punct   7:punct _
6       5       5       NUM     CD      NumType=Card    7       nummod  7:nummod        _
7       Euros   euro    NOUN    NNS     Number=Plur     4       conj    1:appos|4:conj:or       _
8       or      or      CCONJ   CC      _       10      cc      10:cc   _
9       7,5     7,5     NUM     CD      NumType=Card    10      nummod  10:nummod       _
10      Euros   euro    NOUN    NNS     Number=Plur     4       conj    1:appos|4:conj:or       _
11      (       (       PUNCT   -LRB-   _       12      punct   12:punct        SpaceAfter=No
12      1       1       X       LS      _       4       dep     4:dep   SpaceAfter=No
13      )       )       PUNCT   -RRB-   _       12      punct   12:punct        _
14      for     for     ADP     IN      _       18      case    18:case _
15      a       a       DET     DT      Definite=Ind|PronType=Art       18      det     18:det  _
16      72      72      NUM     CD      NumType=Card    17      nummod  17:nummod       _
17      heures  heure   NOUN    NNS     Number=Plur     18      compound        18:compound     _
18      lenght  lenght  NOUN    NN      Number=Sing     4       nmod    4:nmod:for      SpaceAfter=No
19      .       .       PUNCT   .       _       1       punct   1:punct _
```

**Comment by amir-zeldes** [comment]
I would change the comma to a period. Otherwise it's like saying that "7.5" and "7,5" are different lexical items, no? And if we really care if it was comma or period we can use the FORM field.
